Add subprotocol negotiation (#144)
ietf-wg-webtrans/draft-ietf-webtrans-http3: draft-ietf-webtrans-http3.md

HTTP-DATAGRAM}} is not required by WebTransport and can safely be ignored by WebTransport
endpoints.

## Subprotocol Negotiation

WebTransport over HTTP/3 offers a subprotocol negotiation mechanism, similar to
TLS Application-Layer Protocol Negotiation Extension (ALPN) {{?RFC7301}}; the
intent is to simplify porting pre-existing protocols that use QUIC and rely on
this functionality.

The user agent MAY include a `WebTransport-Subprotocols-Available` header field
in the CONNECT request, enumerating the possible subprotocols; if received, the

> server MUST include a `WebTransport-Subprotocol` field in the 2xx response, or

Reviewer: What does the server put in this field?

Author: Clarified.

respond with an error code instead.

Both `WebTransport-Subprotocols-Available` and `WebTransport-Subprotocol` are
Structured Fields {{!RFC8941}}. `WebTransport-Subprotocols-Available` is a List
of Tokens, and `WebTransport-Subprotocol` is a Token. The token in the
`WebTransport-Subprotocol` response header field MUST be one of the tokens
listed in `WebTransport-Subprotocols-Available` of the request.

> 

Reviewer: Are these tokens registered with IANA? We should state the answer in the draft

Author: Clarified.

HTTP-DATAGRAM}} is not required by WebTransport and can safely be ignored by WebTransport
endpoints.

## Subprotocol Negotiation

WebTransport over HTTP/3 offers a subprotocol negotiation mechanism, similar to
TLS Application-Layer Protocol Negotiation Extension (ALPN) {{?RFC7301}}; the
intent is to simplify porting pre-existing protocols that use QUIC and rely on
this functionality.

The user agent MAY include a `WebTransport-Subprotocols-Available` header field

> in the CONNECT request, enumerating the possible subprotocols; if received, the
> server MUST include a `WebTransport-Subprotocol` field with the subprotocol
> chosen in the 2xx response, or respond with an error code instead.

Reviewer: This MUST makes this feature mandatory for servers (as opposed to optional for clients). I'd suggest keeping the optionality, similar to how ALPN works. For example:
```suggestion
in the CONNECT request, enumerating the possible subprotocols. If the server
receives such a header, it MAY include a `WebTransport-Subprotocol` field in
a successful (2xx) response. If it does, the server SHALL include a single
subprotocol from the client's list in that field. Servers MAY reject the request
if the client did not include a suitable subprotocol.
```

Author: Well, it's only mandatory if the client already expects the server to support it.

Reviewer 2: Feels like we ought to say something about the case where the client asked for "foo, bar" and the server says "200" but doesn't include the header.

Both `WebTransport-Subprotocols-Available` and `WebTransport-Subprotocol` are
Structured Fields {{!RFC8941}}. `WebTransport-Subprotocols-Available` is a List
of Tokens, and `WebTransport-Subprotocol` is a Token. The token in the
`WebTransport-Subprotocol` response header field MUST be one of the tokens
listed in `WebTransport-Subprotocols-Available` of the request.  The semantics

> of individual token values is entirely up to the WebTransport resource in

Reviewer 3: ```suggestion
of individual token values is entirely determined by the WebTransport resource in
```

> question.

Reviewer: ```suggestion
question, and are not registered with IANA in the way that ALPN tokens are.
```

Author: This is technically not true. Assume I have a Foobar Protocol, and it has subprotocols `foo` and `bar`. I can create an IANA registry for subprotocols supported by the Foobar resources, and put `foo` and `bar` into it.

Reviewer: ```suggestion
question, and are not registered in IANA's "ALPN Protocol IDs" registry.
```
